polynomial/__init__.py

```diff
@@ -155,27 +155,45 @@ def __str__(self):
         if self.degree < 0:
             return "0"
 
-        def ones_removed(ak, k):
-            #  the  coefficients before the non-zero-degree terms
-            #  should not be explicitly displayed if they are
-            #  1 or -1
-            if ak == 1 and k != 0:
-                return ""
-            if ak == -1 and k != 0:
-                return "-"
-            return ak
-
-        terms = ["{0}x^{1}".
-                 format(ones_removed(ak, k), k)
-                 for k, ak in enumerate(self._vector)
+        # Hacky method to check if a component is the leading term.
+        is_leading = [True]
+
+        def components(ak, k):
+            ak = str(ak)
+
+            if ak[0] == "-":
+                # Strip - from ak
+                ak = ak[1:]
+                sign = "- "
+            else:
+                sign = "+ "
+
+            if is_leading[0]:
+                sign = "-" if sign == "- " else ""
+                is_leading[0] = False
+
+            # if ak is 1, the 1 is implicit when raising x to non-zero k,
+            # so strip it.
+            ak = "" if ak == "1" and k != 0 else ak
+
+            # set x^k portion.
+            if k == 0:
+                p, k = "", ""
+            elif k == 1:
+                p, k = "x", ""
+            else:
+                p = "x^"
+
+            return sign, ak, p, k
+
+        # 0: sign, 1: coeff, 2: x^, 3: a
+        # eg. -         5       x^     2
+        terms = ["{0}{1}{2}{3}".
+                 format(*components(ak, self.degree - k))
```

Comment:
Maybe we could just make `is_leading` the third argument of `components`? So that we call the method here like `format(*components(ak, self.degree - k, k == 0))`. This would be more clear and less 'hacky' in my opinion.

Reply:
I realized this had another issue, in the case that self.degree is incorrect (vector was resized or we lost the leading term and need to move back (eg. Polynomial(1, 2) - Polynomial(1, 0) gives degree 1 instead of 0) - I changed degree to trim the leading 0's off _vector to fix this.

In addition, I noticed a common need for accessing coefficients and degrees, so I decided to add a terms setter and getter, which returns all non-zero terms of the polynomial in tuple form / sets them according. This would be useful for the math operations since we can avoid the 0 elements, and would also be useful if we want to use something aside from _vector.